### ISSUE TITLE
for 0 in ... do gives warning "Incomplete pattern matches" (FS0025)

### ISSUE THREAD

**Issue by abelbraaksma** [dotnet/fsharp#14140]
Ok, to be fair, I understand why this happens, but still, reporting it as I believe we can come up with a better warning in cases like this.

![image](https://user-images.githubusercontent.com/16015770/196536996-5aeece93-544a-4702-8394-55259f7cad56.png)

## Repro steps

Any `for..in..do` loop with a constant:

```f#
for 0 in 1..10 do  // warning on '0'
    ()
```

Or 
```f#
[<Literal>]
let Foo = 42

for Foo in 1..10 do  // warning on Foo
   ()
```

## Expected behavior

A warning, sure, but not this one. It's rather uncommon for people to know, let alone use, patterns in the variable designation of a for-loop.

Maybe something like this:

> Warning: use of a constant in a for-loop identifier position may lead to runtime errors, maybe you meant to use a variable or a pattern?

## Actual behavior

You get a (rather confusing, esp. to newbies) warning:

> warning FS0025: Incomplete pattern matches on this expression. For example, the value '0' may indicate a case not covered by the pattern(s).

## Known workarounds

The code is clearly a mistake, so giving a warning is good.Workaround is obviously to fix the code.

## Related information

As silly as it may sound, this warning actually caught me off-guard. It's probably just late in the day and I wrote `for 0 in 1..10 do`, not immediately realizing it was supposed to be `for _ in 1..10 do`.

Anyway, just throwing it out here to put on the list of warning/errors that could benefit from some attention.

**Comment by kerams** [comment]
> It's rather uncommon for people to know, let alone use, patterns in the variable designation of a for-loop.

I wouldn't go that far. You might see the `KeyValuePair` active pattern in that position or possibly union case unwrapping (be it single case or even `for Ok v in errorsFilteredOut do`, in which case FS0025 is very appropriate).

> Just tried the following, never even considered it, but this is legal (again, I understand why it is legal, but still, this is just weird!)

Yeah, you should be able to do a similar thing at every `SynPat` and `SynSimplePat`.

```fsharp
[] |> List.map (fun 0 -> 0)
```

So the problem is, how do we decide whether to emit FS0025 or something else? When the pattern is a constant in anything but a match clause?

**Comment by abelbraaksma** [comment]
> When the pattern is a constant in anything but a match clause?

@kerams, tx for chiming in! That seems reasonable. In `match .. with` and `function | 0 | 1 -> ...` it is reasonable to use constants. In a `let 0 = 10` pattern not so much. I'd suggest the same error _number_, just with a different message.

**Comment by charlesroddie** [comment]
I think this should go in FsLang discussion, or a slack discussion, as the feature request here is not clear at all, or the scope of it. In the worst case someone might attempt to fix this by adding a special case, thereby adding complexity to the language.

My suggestion for this would be that there should be a concept of a SimpleGuaranteedPattern which includes `x`, `(x, y)`, but excludes `Some x`, `0`, `"z"`, active patterns. These deconstruct using only a single pattern, reliably.

Then `for [pat] in ...`, `fun [pat] ->` would only work for SimpleGuaranteedPatterns.
`function ...` and `match` would not require this as they allow multiple patterns.

This would be a breaking change.

**Comment by abelbraaksma** [comment]
@charlesroddie I didn't want to go so far as to change the language, though your suggestions are good. The reason I raised this is that the warning isn't very useful to newbies, and sometimes seasoned programmers alike and I merely suggested to improve it. Not to forbid it, which would be a breaking change.

**Comment by charlesroddie** [comment]
Without a change to the language, "incomplete pattern matches on this expression" is the _correct_ reason for the error.

> Warning: use of a constant in a for-loop identifier position may lead to runtime errors

"Use of an incomplete pattern may lead to runtime errors" is better as the issue applies to more than constants, and also does not apply to the constant `()`. And this error is pretty much what we have now.

If you correct to "Use of an constant non-unit pattern may lead to runtime errors" then you just split up a concept into multiple examples unnecessarily, increasing complexity.

> maybe you meant to use a variable or a pattern?

I think this goes down a long road that should be separate from error messages. I believe that error messages should point out the error - it's self-contained and helpful to do that. Any guesses as to how the user might improve the code, or what the user might be thinking, whether AI-based or algorithmic, can be added by IDE features.

**Comment by abelbraaksma** [comment]
For a long time we’re on the road of improving error messages. For instance, when an overload cannot be found, we show suggestions.

One of the major complaints about newcomers coming to the language is vague and seemingly unrelated error messages, like the one in the example.

Sure, my proposed wording may be a little too specific, but surely you wouldn’t disagree that we should improve the coding experience by providing better, more humanly readable and understandable errors? If that means having a little match statement that improves it for two or three designated cases, I don’t see how that really increases the overall complexity of the F# code base.

**Comment by vzarytovskii** [comment]
Oh yeah, this is very confusing message. We should definitely improve it.

**Comment by charlesroddie** [comment]
> Sure, my proposed wording may be a little too specific, but surely you wouldn’t disagree that we should improve the coding experience by providing better, more humanly readable and understandable errors? If that means having a little match statement that improves it for two or three designated cases, I don’t see how that really increases the overall complexity of the F# code base.

Yes it can be improved but:
- The current suggestion doesn't improve it and introduces conceptual unclarity and 2 mistakes.
- I have the impression that there is almost no review process for code that changes error messages, so there is a risk of prematurely merging PRs here without first having concrete suggestions, identifying the underlying issue, and with the possibility of improving certain areas and making others worse.

In this case the alternative suggestion has false positives (since constants are fine sometimes), and does not cover all cases (non-constant incomplete patterns), suggesting it is not the best way to explain the issue to users. It also strongly implies it may lead to runtime exceptions, which is not true. `for 42 in 1..10 do ()` does nothing, just like `for _ in 1 .. 10`, as the current warning clearly implies (on my system - it's not in your screenshot): `Unmatched elements will be ignored.`

I can create a language suggestion that would improve this (with deprecation or breaking changes), and that might inform what a more helpful error message would look like even if these changes aren't accepted.

**Comment by abelbraaksma** [comment]
> The current suggestion doesn't improve it and introduces conceptual unclarity and 2 mistakes.

Absolutely, it was not a proposal, just a hint as to what direction I was aiming for. I thought the phrase “something like this” made that clear.

> I can create a language suggestion that would improve this (with deprecation or breaking changes)

I assume you meant “without” instead of “with”?

All we need is an improvement over the current error. I agree it shouldn’t create false positives or negatives, but changing a warning _message_ has never led to breaking changes. If you can think of a better message (I’ll try too), by all means, go ahead and write it in this thread.

Luckily, we don’t need language suggestions for changing warnings, that’d be a bit much. If you’ve got a good text, or someone else, I’d be happy to make the change.

**Comment by dsyme** [comment]
Yes this is not a bug, but the error message could be improved - actually also at argument positions and other places where incomplete patterns really aren't expected, e.g.

```fsharp
> let f 0 = 1;;

  let f 0 = 1;;
  ------^

stdin(2,7): warning FS0025: Incomplete pattern matches on this expression. For example, the value '1' may indicate a case not covered by the pattern(s).

val f: int -> int
```


Agreed with @charlesroddie that the proposed diagnostic needs more work. Possibly

```
Incomplete pattern matches on this expression. Did you use a constant where a loop variable was expected?  For example, the value '1' may indicate a case not covered by the pattern(s).

Incomplete pattern matches on this expression. Did you use a constant where an argument name was expected?  For example, the value '1' may indicate a case not covered by the pattern(s).
```

**Comment by abelbraaksma** [comment]
Sounds like a much cleaner warning message, agreed.
